transport/cluster.go

```diff
@@ -0,0 +1,288 @@
+package transport
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"strconv"
+	"time"
+
+	"scylla-go-driver/frame"
+	. "scylla-go-driver/frame/response"
+
+	"go.uber.org/atomic"
+)
+
+type PeerMap = map[string]*Node
+
+type refreshHandler chan struct{}
+
+type Cluster struct {
+	peers         atomic.Value // PeerMap
+	control       *Conn
+	cfg           ConnConfig
+	events        responseHandler // TODO: should events have their own type of chan or is responseHandler fine?
+	handledEvents []frame.EventType
+	knownHosts    []string // TODO: do we want to update known hosts with node info from cluster?
+	refresher     refreshHandler
+
+	// Yet to be used.
+	// ring        map[Token]*Node
+	// dataCenters map[string]*DataCenter
+}
+
+// type dataCenter struct {
+// 	nodes   []Node
+// 	rackCnt uint // Type?
+// }
+
+// CreateCluster also creates control connection and starts handling events and refreshing topology.
+func CreateCluster(cfg ConnConfig, e []frame.EventType, hosts ...string) (*Cluster, error) {
+	if len(hosts) < 1 {
+		return nil, fmt.Errorf("at least one host is required to create cluster")
+	}
+
+	p := atomic.Value{}
+	p.Store(PeerMap{})
+	k := make([]string, len(hosts))
+	copy(k, hosts)
+
+	c := &Cluster{
+		peers:         p,
+		cfg:           cfg,
+		handledEvents: e,
+		knownHosts:    k,
+	}
+
+	if err := c.OpenControl(); err != nil {
+		return nil, fmt.Errorf("creating cluster: %w", err)
+	}
+	if err := c.refreshTopology(); err != nil {
+		return nil, fmt.Errorf("creating cluster: %w", err)
+	}
+
+	go c.handleEvents()
+	go c.loop()
+	return c, nil
+}
+
+func (c *Cluster) OpenControl() error {
+	// TODO: should we random shuffle known hosts after establishing connection?
```

Comment:
I don't find it helpful.

Reply:
I found something similar in rust so I wanted to have second opinion on this.